packages/flutter/lib/src/semantics/semantics.dart

```diff
@@ -4400,10 +4626,51 @@ class SemanticsOwner extends ChangeNotifier {
         }
       }
     }
+
     visitedNodes.sort((SemanticsNode a, SemanticsNode b) => a.depth - b.depth);
     final SemanticsUpdateBuilder builder = SemanticsBinding.instance.createSemanticsUpdateBuilder();
+
+    final List<SemanticsNode> updatedVisitedNodes = <SemanticsNode>[];
+
     for (final SemanticsNode node in visitedNodes) {
-      assert(node.parent?._dirty != true); // could be null (no parent) or false (not dirty)
+      final bool isTraversalParent = node._isTraversalParent;
+      final bool isTraversalChild = node._isTraversalChild;
+
+      if (kIsWeb) {
+        updatedVisitedNodes.add(node);
+      } else {
+        if (!isTraversalParent && !isTraversalChild) {
+          updatedVisitedNodes.add(node);
+          continue;
+        }
+
+        if (isTraversalChild) {
+          // If the node is a child of an overlay portal, we add its parent
+          // node in the `updatedVisitedNodes` list for later grafting to generate
+          // a correct traversal order.
+          final SemanticsNode? parentNode = _traversalParentNodes[node.traversalChildIdentifier];
+          if (parentNode != null && !updatedVisitedNodes.contains(parentNode)) {
+            updatedVisitedNodes.add(parentNode);
+          }
+        }
+
+        updatedVisitedNodes.add(node);
+      }
+
+      // If the node is an overlay portal parent, then add it to the
+      // _traversalParentNodes map. Similarly, add the node to the
+      // _traversalChildNodes map if it is an overlay portal child.
+      if (isTraversalParent) {
+        _traversalParentNodes[node.traversalParentIdentifier!] = node;
+      } else if (isTraversalChild) {
+        _traversalChildNodes[node.traversalChildIdentifier!] = node;
```

Comment:
when do we clean up the _traversalChildNodes and _traversalParentNodes when nodes in it are disposed